presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -594,6 +595,8 @@ private PlanWithProperties planTableScan(TableScanNode node, Expression predicat
                             .map(node.getAssignments()::get)
                             .collect(toImmutableSet())));
 
+            layouts = filteredCopy(layouts, layout -> !layout.getLayout().getPredicate().isNone());
```

Comment:
When would you get back a layout where `layout.getLayout().getPredicate().isNone()` vs. when would you get an empty list from the start?

Reply:
See: com.facebook.presto.metadata.MetadataManager#getLayouts, you get empty layouts when you put into a `Constraint` a `TupleDomain.none()`.  
`layout.getLayout().getPredicate().isNone()` depends on connector. For Hive you will get predicate produced by `com.facebook.presto.hive.HiveMetadata#createPredicate`